### ISSUE TITLE
TSQL false positive RF05 + AL05 on table-valued method alias; sqlfluff fix causes script corruption

### ISSUE THREAD

**Issue by exit81** [sqlfluff/sqlfluff#6733]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The following select statement is valid TSQL using the table-valued method [nodes()](https://learn.microsoft.com/en-us/sql/t-sql/xml/nodes-method-xml-data-type?view=sql-server-ver16):
```
select r.c.value('column[1]', 'varchar(10)') as 'Value'
from @xml.nodes('/rows/row') as r(c);
```

 `sqlfluff lint` incorrectly raises RF05 (do not use special characters in identifiers) on the column expression and AL05 (alias is never used in SELECT statement) on the method expression. The identifier does not contain special characters, and the alias is used. `sqlfluff fix` will delete the alias, which results in invalid TSQL. Executing the "fixed" script will produce the following error message:
```
Msg 318, Level 15, State 0, Line 4
The table (and its columns) returned by a table-valued method need to be aliased.
```

Additionally, `sqlfluff lint` raises LT01 (expected single whitespace between naked identifier and start bracket) on the alias. Microsoft's docs do not have spaces in the aliases, e.g.: `Instructions.nodes('/root/Location') AS T2(Loc)`

### Expected Behaviour

`.` accessor is appropriately recognized and not considered a special character.
Usage of table-valued method alias is appropriately recognized.

### Observed Behaviour

`.` accessor is treated as a special character in the identifier.
Usage of table-valued method alias is not recognized.

### How to reproduce

1. Create `xml.sql`
```
-- xml.sql
declare @xml xml = '<rows><row><column>asdf</column></row></rows>';

select r.c.value('column[1]', 'varchar(10)') as 'Value'
from @xml.nodes('/rows/row') as r(c);
```

2. Perform `sqlfluff lint xml.sql` and see the spurious RF05, AL05 and LT01 violations.
```
== [xml.sql] FAIL
L:   4 | P:  49 | RF05 | Do not use special characters in identifiers.
                       | [references.special_chars]
L:   5 | P:  33 | AL05 | Alias 'r' is never used in SELECT statement.
                       | [aliasing.unused]
L:   5 | P:  34 | LT01 | Expected single whitespace between naked identifier and
                       | start bracket '('. [layout.spacing]
```

3. Perform `sqlfluff fix xml.sql` and see that the alias has been deleted:
```
-- xml.sql
declare @xml xml = '<rows><row><column>asdf</column></row></rows>';

select r.c.value('column[1]', 'varchar(10)') as 'Value'
from @xml.nodes('/rows/row');
```

4. Execute the script in SQL Server and see the error message:
```
Msg 318, Level 15, State 0, Line 4
The table (and its columns) returned by a table-valued method need to be aliased.
```

### Dialect

TSQL

### Version

sqlfluff, version 3.3.1
Python 3.12.9

### Configuration

```
[sqlfluff]
dialect = tsql
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by msergein** [comment]
I also encountered xml parser errors for tsql syntax when alias is specified:

```sql
DECLARE @Xml XML = '<a>1</a>';

-- (1) parses cleanly
SELECT 1 FROM @Xml.nodes('a');

-- (2) PRS
SELECT 1 FROM @Xml.nodes('a') AS FN;

-- (3) PRS
SELECT 1 FROM @Xml.nodes('a') AS FN(a);
```

Output for (2):
```
Found unparsable section: 'AS FN;'
```

Output for (3)
```
Found unparsable section: 'AS FN(a);'
```

sqlfluff, version 4.1.0
Python 3.14.4 (Windows)

**Comment by vini-haa** [comment]
I'd like to work on this — will open a PR fixing the TSQL parsing of `@xml.nodes(...) AS r(c)` aliases.

**Comment by Pawansingh3889** [comment]
Took a look on current `main`. AL05 no longer reproduces here — only RF05 and LT01 still fire:

```
L3 P49  RF05  on the alias  quoted_identifier 'Value'
L4 P34  LT01  on the '(' of the  r(c)  column-list alias
```

The `as r(c)` form actually parses fine (`alias_expression` -> `naked_identifier` + bracketed `identifier_list`), so this is rule behaviour rather than a parse gap.

The LT01 case looks clear-cut: T-SQL column-list aliases are written without a space between the alias and the bracket (`T2(Loc)` in the MS docs), so requiring whitespace there is wrong. Happy to put a PR up for that one.

RF05 on the single-quoted alias is more debatable — `'Value'` string-literal aliases are discouraged in T-SQL anyway (MS recommends `[Value]` or `"Value"`), so flagging it isn't clearly a false positive. I'd split that out from the LT01 fix.

**Comment by Pawansingh3889** [comment]
Retested this on 4.2.1 with the exact repro — an update for the thread, since it has moved:

**The corruption is fixed.** The parse tree is now correct (`r.c.value(...)` and `@xml.nodes(...)` both parse as `datatype_method`, and `as r(c)` parses as an alias with an `identifier_list`), AL05 no longer fires, and `sqlfluff fix` keeps the alias — the output is valid T-SQL.

**Two false positives remain:**

1. RF05 still flags the `'Value'` alias. T-SQL allows single-quoted aliases, and the segment parses as `quoted_identifier`, so it looks like RF05's quote handling recognises `"..."` and `[...]` but not the T-SQL `'...'` form — the apostrophes are being counted as special characters.
2. LT01 still wants a space in `as r(c)` and "fixes" it to `as r (c)` — valid, but the opposite of the style in Microsoft's own `nodes()` documentation. The alias-to-column-list bracket presumably needs touch spacing in the T-SQL dialect config.

I'd like to take both — RF05 first since it's the misleading one. Will open separate PRs unless anyone objects.

**Comment by adhavan18** [comment]
@Pawansingh3889 are you still planning to take these two? Your retest above is the useful part of this thread and I do not want to duplicate it.

Asking because it has been about a month and neither has appeared yet. If you have moved on I would pick up the RF05 half, since a single-quoted T-SQL alias parsing as `quoted_identifier` and still being flagged looks like a straightforward gap in the rule's quote handling rather than a design question. If you are still on it, I will leave both alone.

**Comment by adhavan18** [comment]
retested both halves on current main (`a5ad86d`) before picking anything up, and they have diverged since @Pawansingh3889's 4.2.1 run:

**RF05 no longer reproduces.** the exact repro is clean:

```
$ sqlfluff lint t.sql --dialect tsql --rules RF05
All Finished!
```

the quote handling is fine now. `raw_normalized()` strips all three t-sql forms identically, so there is nothing left for the special-chars check to trip on:

```
'SELECT 1 AS [Value]'   -> raw='[Value]'  normalized='Value'
'SELECT 1 AS "Value"'   -> raw='"Value"'  normalized='Value'
"SELECT 1 AS 'Value'"   -> raw="'Value'"  normalized='Value'
```

so the apostrophes are not being counted as special characters any more.

**LT01 still reproduces**, exactly as described:

```
L:   3 | P:  26 | LT01 | Expected single whitespace between naked identifier and ...
$ sqlfluff fix --rules LT01
FROM @Xml.nodes('a') AS r (c);
```

which is the opposite of how microsoft writes it in the `nodes()` docs.

one note on the shape of a fix, in case it saves someone time: the alias parses as `alias_expression -> naked_identifier -> bracketed`, and `bracketed` is a generic type. so it cannot be fixed with a `[sqlfluff:layout:type:bracketed]` entry without changing spacing for every bracketed expression everywhere. it needs the alias's column list to carry its own segment type first, then a layout rule keyed to that.

@Pawansingh3889 that leaves only the LT01 half of what you claimed, so if you are still on it the ground is clearer than it was. i am not picking it up. if you have moved on, say so and i will take the typed-segment approach above.

**Comment by Pawansingh3889** [comment]
@adhavan18 sorry for the delay, and thanks for checking rather than duplicating.

Both halves are now covered, so please don't spend time on either:

- **RF05**: fixed in #8088 (merged). Root cause was a level below the rule. The tsql dialect patches the `single_quote` lexer for `''` escapes, and the patch dropped the `segment_kwargs` the ansi lexer carries, so `raw_normalized()` returned the value with its delimiters attached. Your read was right that it was a gap in quote handling rather than a design question.
- **LT01**: just opened #8285.

One correction to what I wrote above, since it shaped how I framed the LT01 half. I said "requiring whitespace there is wrong," citing `T2(Loc)` in the docs. The first part holds up, since every SQL example in the [`nodes()` reference](https://learn.microsoft.com/en-us/sql/t-sql/xml/nodes-method-xml-data-type) attaches the bracket (`AS T2(Loc)`, `T(c)`, `A2(B)`, `AS T1(Locations)`). But the [FROM syntax notation](https://learn.microsoft.com/en-us/sql/t-sql/queries/from-transact-sql) writes it spaced, as `[ ( column_alias [ , ...n ] ) ]`, so the docs don't actually mandate the touching form the way I implied.

That changed the fix. Forcing `touch` would have fixed the false positive but flipped the mandate, and `sqlfluff fix` would then rewrite the spaced form the rule has been requiring until now, which is churn beyond the reported bug. So the PR wraps the column list in an `alias_column_list` segment (it previously parsed as a bare `bracketed`, which layout config can't target) and sets `spacing_before = any`: neither spacing is enforced, and either can be required by configuring that segment.

If you think the opinionated default is the better call, that's a one-line change and worth arguing on the PR. I don't feel strongly, I just didn't want to rewrite valid SQL on people's behalf without a clear mandate for it.

**Comment by Pawansingh3889** [comment]
@adhavan18 our comments crossed, so mine above reads like I ignored yours. I hadn't seen it when I wrote it, and credit where it's due: you landed on the same fix independently.

Your note that it "needs the alias's column list to carry its own segment type first, then a layout rule keyed to that" is exactly what #8285 does. The new segment is `alias_column_list`, and the layout entry is keyed to it. You also saved me from second-guessing the RF05 half, since your retest on `a5ad86d` confirming `raw_normalized()` now strips all three quote forms identically is a cleaner check on #8088 than the one I ran.

The one place I went further than your description is the value. I set `spacing_before = any` rather than requiring the touching form, for the reason in my previous comment: the spaced form is what the rule has been demanding until now, so mandating the opposite would have `sqlfluff fix` rewriting files that are currently clean.

If you have any appetite for reviewing it, you have more context on this than anyone.

**Comment by be-student** [comment]
The RF05 and AL05 findings are fixed on current `main`, but LT01 still reproduces. I’m working on that remaining false positive with a regression test.
